File: 2024/day_6/day6.py
# Advent of Code - Day 6 | Antonin Censier (version Python)

def lire_grille(fichier):
    """Lit la grille depuis data.txt et renvoie (grille, x, y, direction)."""
    grille = [list(line.strip()) for line in open(fichier)]
    start_x = start_y = None

    for y, ligne in enumerate(grille):
        for x, c in enumerate(ligne):
            if c == '^':  # départ toujours vers le haut
                start_x, start_y = x, y
                grille[y][x] = '.'  # on remplace le garde par une case vide
                return grille, start_x, start_y, 0  # 0 = direction haut
    raise ValueError("Pas de garde trouvé (^)")


def in_bounds(x, y, grille):
    """Retourne True si (x, y) est dans la grille."""
    return 0 <= y < len(grille) and 0 <= x < len(grille[0])


def tourne_droite(direction):
    """Renvoie la nouvelle direction après un virage à droite."""
    return (direction + 1) % 4


def parcourir(grille, x, y, direction):
    """Fait avancer le garde jusqu'à sortir de la grille."""
    # directions : 0 = haut, 1 = droite, 2 = bas, 3 = gauche
    # Format: (dx, dy) pour cohérence avec x, y
    mouvements = [(0, -1), (1, 0), (0, 1), (-1, 0)]
    
    while True:
        # Marque la case actuelle comme visitée
        grille[y][x] = 'X'
        
        dx, dy = mouvements[direction]
        nx, ny = x + dx, y + dy
        
        # Si on sort de la grille → arrêt
        if not in_bounds(nx, ny, grille):
            break
        
        # Si mur → tourne à droite
        if grille[ny][nx] == '#':
            direction = tourne_droite(direction)
        else:
            # avance normalement
            x, y = nx, ny
    
    return grille


def compter_X(grille):
    """Compte le nombre de cases 'X' visitées."""
    return sum(ligne.count('X') for ligne in grille)


def afficher_grille(grille):
    """Affiche la grille dans la console."""
    for ligne in grille:
        print(''.join(ligne))
    print()


def main():
    fichier = "data.txt"
    grille, x, y, direction = lire_grille(fichier)
    grille = parcourir(grille, x, y, direction)
    afficher_grille(grille)
    resultat = compter_X(grille)
    print("Résultat partie 1 :", resultat)


if __name__ == "__main__":
    main()

File: 2024/day_6/day_6.cpp
/* Advent of Code - Day 6 | Antonin Censier */
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <string>
#include <limits>

class Curseur {
    public : 
        int x, y;
        char direction;
        Curseur(int x, int y) : x(x), y(y){
            std::cout<<"x = "<<x<<"y = "<<y<<std::endl;
            direction = 'h';
        };

        void parcour(std::vector<std::vector<char>>& grille);
        bool outOfBound(std::vector<std::vector<char>>& grille);

};

void Curseur::parcour(std::vector<std::vector<char>>& grille) {
    while (true) {
        // calculer le prochain pas
        int nx = x, ny = y;
        if (direction == 'h') ny--;
        else if (direction == 'd') nx++;
        else if (direction == 'b') ny++;
        else if (direction == 'g') nx--;
    
        // si on sort, stop
        if (ny < 0 || nx < 0 || ny >= (int)grille.size() || nx >= (int)grille[0].size()) {
            grille[y][x] = 'X'; // marque la dernière case encore dans la grille
            break;
        }
    
        // si mur, tourne à droite
        if (grille[ny][nx] == '#') {
            if (direction == 'h') direction = 'd';
            else if (direction == 'd') direction = 'b';
            else if (direction == 'b') direction = 'g';
            else if (direction == 'g') direction = 'h';
        } else {
            grille[y][x] = 'X'; // marque la case actuelle
            x = nx;
            y = ny;
        }
    }
    
}

bool Curseur::outOfBound(std::vector<std::vector<char>>& grille){
    return (y < 0 || x < 0 || y >= grille.size() || x >= grille[0].size());
}

Curseur getGrille(std::vector<std::vector<char>>& grille) {

    std::ifstream file;
    file.open("data.txt");

    std::string line;
    int count_l = 0;

    Curseur cursed(-1, -1);
    int count_c = 0;

    while(std::getline(file, line)){     
        std::vector<char> thisline;

        if(count_c != 0){
            thisline.reserve(count_c);
        }

        for (count_c = 0; count_c < line.size(); ++count_c) {
            char c = line[count_c];
            if (c == '^') {
                cursed = Curseur(count_c, count_l);
            }
            thisline.push_back(c);
        }   

        grille.push_back(thisline);
        count_l++;
    }
    return cursed;
}

void afficheGrille(std::vector<std::vector<char>>& grille){
    for(std::vector<char> l : grille){
        for(char c : l){
            std::cout<<c;
        }
        std::cout<<std::endl;
    }
}

int getX(std::vector<std::vector<char>>& grille){
    int compte = 0;
    for(std::vector<char> l : grille){
        for(char c : l){
            if(c == 'X') compte++;
        }
    }
    return compte;
}

int main() {
    std::vector<std::vector<char>> grille;
    Curseur initPos = getGrille(grille);
    afficheGrille(grille);
    std::cout<< "check the init pos : "<<grille[initPos.y][initPos.x]<<std::endl;
    initPos.parcour(grille);
    afficheGrille(grille);
    int res = getX(grille);
    std::cout<<"Resultat partie 1"<<res<<std::endl;
}